src/lib/transition_router/initial_validator.ml

```diff
@@ -196,39 +196,53 @@ let run ~logger ~trust_system ~verifier ~transition_reader
              Duplicate_block_detector.check ~precomputed_values
                duplicate_checker logger transition_with_hash ;
              let sender = Envelope.Incoming.sender transition_env in
-             let defer f = Fn.compose Deferred.return f in
-             match%bind
-               let open Deferred.Result.Monad_infix in
-               External_transition.(
-                 Validation.wrap transition_with_hash
-                 |> defer
-                      (validate_time_received ~precomputed_values
-                         ~time_received)
-                 >>= defer
-                       (validate_genesis_protocol_state ~genesis_state_hash)
-                 >>= (fun x -> validate_proofs ~verifier [x] >>| List.hd_exn)
-                 >>= defer validate_delta_transition_chain
-                 >>= defer validate_protocol_versions)
-             with
-             | Ok verified_transition ->
-                 External_transition.poke_validation_callback
-                   (Envelope.Incoming.data transition_env)
-                   valid_cb ;
-                 Envelope.Incoming.wrap ~data:verified_transition ~sender
-                 |> Writer.write valid_transition_writer ;
-                 let blockchain_length =
-                   External_transition.Initial_validated.consensus_state
-                     verified_transition
-                   |> Consensus.Data.Consensus_state.blockchain_length
-                   |> Coda_numbers.Length.to_int
-                 in
-                 Coda_metrics.Transition_frontier
-                 .update_max_blocklength_observed blockchain_length ;
-                 return ()
-             | Error error ->
-                 Coda_net2.Validation_callback.fire_exn valid_cb `Reject ;
-                 handle_validation_error ~logger ~trust_system ~sender
-                   ~state_hash:(With_hash.hash transition_with_hash)
-                   ~delta:genesis_constants.protocol.delta error )
+             let computation =
+               let open Interruptible.Let_syntax in
+               let defer f x =
+                 Interruptible.uninterruptible @@ Deferred.return (f x)
+               in
+               let%bind () =
+                 Interruptible.lift Deferred.unit
+                   (Coda_net2.Validation_callback.await valid_cb)
```

Comment:
It might make more sense to add an `await_timeout` method for this, since we don't want the callback being called to trigger the interrupt. In the `Error`branch below, I think the current behaviour may prevent the `handle_validation_error` from being called.